### PR TITLE
Document the process anti pattern of sending large data

### PR DESCRIPTION
**Pull request by PragTob**
As per usual, thanks to all of ya for your great work!

Follow up to/extension of #13173

First draft, happy to adjust and extend it in many ways. I had trouble coming up with a simple example as we needed a bunch of data to make sure it's not good. I could have employed the first anti pattern myself and did repeated statistics calculation, but that'd have been worse :sweat-smile:

I remembered José's comment around sending along the `conn` and figured it's central enough in elixir to not throw anyone off.

If someone has a better example, happy to redo it!

Thanks y'all! :green-heart:

### CODE REVIEW

**Comment by josevalim** [comment]
The original anti-patterns mentioned exactly this, sending data between processes, but it was removed because it is hard to precisely say when you are sending large data. Sometimes you do have to send it and there isn't much we can do.

Therefore, can you please rephrase this as accidentally capturing large data on spawn? I.e. without the sending considerations? Thank you.

**Comment by PragTob** [comment]
@josevalim :wave: Thanks for the context, I was unaware of the previous version.

Rephrasing to validate my understanding:

* Rewrite section to focus on `spawn`/`Task.async` aka where the function captures the large data (to be fully clear, should I write it with `spawn` or `Task.async` as I think the latter is more common)
* Remove mention of this problem around `send` as it's ambiguous what large data is and when it's needed/vs. not needed

Happy to do that! Question though, shouldn't we at least mention `send/2`? While most people will learn about the copying behavior on their elixir/erlang journey I couldn't find it in the elixir docs (at least where I checked, `Process`, `send/2` and the `Process` getting started mentions it's "is stored in the process mailbox" which could be more explicit, however a beginner guide is probably not the best space for this). I don't mean the mention of the "anti-pattern" but that messages are deep copied to the receiving process. Wouldn't want to litter everything, but maybe `Process` docs?

**Comment by josevalim** [comment]
Hrm. It is fine to mention `send` or even `GenServer.call`, but the framing matters. The issue not really sending large data, but sending unnecessary data. Don't send or pass as argument to GenServer.call/GenServer.start_link a large data structure, only to pluck part of it from the server. Pluck the parts that you need on the caller/client.

**Comment by PragTob** [comment]
Makes sense, thanks for the clarification! I'll get back to it later today!

**Comment by PragTob** [comment]
Thanks for the feedback and guidance! New version is up, hope that that's the direction you envisioned. Happy to adjust and iterate in any case :) :green_heart: 


IMG_20170521_203155

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 1]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
 iex> Counter.get(Counter) # After the restart, this process can be used again
 0
 ```
+
+## Sending unnecessary data
+
+#### Problem
+
+Sending a message to a process can be an expensive operation if it is big enough, as messages will be fully copied to the receiving process, which is both CPU and memory intensive. This is due to erlang's "shared nothing" architecture where each process has its own memory, simplifying and speeding up garbage collection.
```
Comment (Nezteb) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:298]:
```suggestion
Sending a message to a process can be an expensive operation if it is big enough, as messages will be fully copied to the receiving process, which is both CPU and memory intensive. This is due to Erlang's "share nothing" architecture where each process has its own memory, simplifying and speeding up garbage collection.
```

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 2]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
...
+That this happens is more obvious when using `send/2`, `GenServer.call/3` or the initial data in `GenServer.start_link/3`, however notably this also happens when using `spawn/1`, `Task.async/1`, `Task.async_stream/3` & friends. It is more subtle here as the anonymous function passed to these captures the variables it references in its closure - meaning that data will also be copied over. Hence, you can accidentally send way more data to a process than you actually need.
```
Comment (Nezteb) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:299]:
```suggestion
This is more obvious when using `send/2`, `GenServer.call/3`, or the initial data in `GenServer.start_link/3`. Notably this also happens when using `spawn/1`, `Task.async/1`, `Task.async_stream/3`, etc. It is more subtle here as the anonymous function passed to these functions captures the variables it references in its closure; this means that all data will be copied over. By doing this, you can accidentally send way more data to a process than you actually need.
```

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 3]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
...
+
+#### Example
+
+To depict the problem let's imagine you were to implement some simple reporting of ip addresses that made requests against your application. You want to do this asynchronously to not block processing, so you decide to use `spawn/1`. It may seem like a good idea to hand over the whole connection ("We might need more data later!"), however it results in copying a lot of unnecessary data (request body, params etc.).
```
Comment (Nezteb) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:303]:
```suggestion
To depict the problem, let's imagine you were to implement some simple reporting of IP addresses that made requests against your application. You want to do this asynchronously and not block processing, so you decide to use `spawn/1`. It may seem like a good idea to hand over the whole connection because we might need more data later. However passing the connection results in copying a lot of unnecessary data like the request body, params, etc.
```

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 4]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
...
+spawn(fn -> log_request_ip(conn.remote_ip) end)
+```
+
+This will still copy over all of `conn`.
+
+#### Refactoring
+
+This anti-pattern has many potential remedies:
+
+* Limiting the data you send to the absolute necessary minimum, instead of just sending the whole struct. For example, don't send an entire `Plug.Conn` struct if all you need is a couple of fields.
```
Comment (Nezteb) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:322]:
```suggestion
* Limit the data you send to the absolute necessary minimum instead of sending an entire struct. For example, don't send an entire `Plug.Conn` struct if all you need is a couple of fields.
```

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 5]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
...
+* If only the process you send the data to needs it, it may fetch the data itself instead.
```
Comment (Nezteb) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:323]:
```suggestion
* If only the process that needs data is the one you are sending to, consider making the process fetch that data instead of passing it.
```

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 6]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
...
+* There are some data structures that are shared between processes and hence don't need copying, such as [ets](https://www.erlang.org/doc/man/ets) and [persistent_term](https://www.erlang.org/doc/man/persistent_term.html).
```
Comment (Nezteb) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:324]:
```suggestion
* There are some data structures that are already shared between processes and don't need copying, such as [ETS](https://www.erlang.org/doc/man/ets) and [`:persistent_term`](https://www.erlang.org/doc/man/persistent_term.html).
```

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 7]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
...
+
+In our case the first, and most common, strategy is applicable. If all we need _right now_ is the ip address, then let's only work with that and make sure that's all we're passing into the closure:
```
Comment (Nezteb) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:326]:
```suggestion
In our case, limiting the input data is a reasonable strategy. If all we need _right now_ is the IP address, then let's only work with that and make sure we're only passing the IP address into the closure, like so:
```

**Review by Nezteb — APPROVED** [review]
Some minor capitalization/wording suggestions. Regardless, I love this content!

**Comment by PragTob** [comment]
@Nezteb thanks for taking the time to review & make suggestions!

IMG_20221012_141750

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 8]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
 iex> Counter.get(Counter) # After the restart, this process can be used again
 0
 ```
+
+## Sending unnecessary data
+
+#### Problem
+
+Sending a message to a process can be an expensive operation if it is big enough, as messages will be fully copied to the receiving process, which is both CPU and memory intensive. This is due to Erlang's "share nothing" architecture where each process has its own memory, simplifying and speeding up garbage collection.
```
Comment (whatyouhide) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:298]: Rephrased to have shorted sentences (easier to read) + less "it"s so that it's easier to follow.

```suggestion
Sending a message to a process can be an expensive operation if the message is big enough. That's because that message will be fully copied to the receiving process, which is both CPU and memory intensive. This is due to Erlang's "share nothing" architecture, where each process has its own memory, which simplifies and speeds up garbage collection.
```

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 9]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
...
+This is more obvious when using `send/2`, `GenServer.call/3`, or the initial data in `GenServer.start_link/3`. Notably this also happens when using `spawn/1`, `Task.async/1`, `Task.async_stream/3`, etc. It is more subtle here as the anonymous function passed to these functions captures the variables it references in its closure; this means that all data will be copied over. By doing this, you can accidentally send way more data to a process than you actually need.
```
Comment (whatyouhide) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:299]:
```suggestion
This is more obvious when using `send/2`, `GenServer.call/3`, or the initial data in `GenServer.start_link/3`. Notably this also happens when using `spawn/1`, `Task.async/1`, `Task.async_stream/3`, and so on. It is more subtle here as the anonymous function passed to these functions captures the variables it references in its closure; this means that all data will be copied over. By doing this, you can accidentally send way more data to a process than you actually need.
```

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 10]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
...
+spawn(fn -> log_request_ip(conn) end)
+```
+
+Please note that this problem also occurs if you think you may just be accessing the relevant parts:
+
+```elixir
+spawn(fn -> log_request_ip(conn.remote_ip) end)
+```
+
+This will still copy over all of `conn`.
```
Comment (whatyouhide) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:316]: More context. Thoughts?

```suggestion
This will still copy over all of `conn`, because the `conn` variable is being captured as a closure inside the spawned function. That function then extracts the `remote_ip` field, but only after the whole `conn` has been copied over.
```

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md [hunk 11]
```diff
@@ -290,3 +290,42 @@ iex> Supervisor.restart_child(App.Supervisor, Counter)
...
+
+#### Refactoring
+
+This anti-pattern has many potential remedies:
+
+* Limit the data you send to the absolute necessary minimum instead of sending an entire struct. For example, don't send an entire `Plug.Conn` struct if all you need is a couple of fields.
+* If only the process that needs data is the one you are sending to, consider making the process fetch that data instead of passing it.
+* There are some data structures that are already shared between processes and don't need copying, such as [ETS](https://www.erlang.org/doc/man/ets) and [`:persistent_term`](https://www.erlang.org/doc/man/persistent_term.html).
+
+In our case, limiting the input data is a reasonable strategy. If all we need _right now_ is the IP address, then let's only work with that and make sure we're only passing the IP address into the closure, like so:
```
Comment (whatyouhide) [inline lib/elixir/pages/anti-patterns/process-anti-patterns.md:326]: I think we usually go with asterisks for italic

```suggestion
In our case, limiting the input data is a reasonable strategy. If all we need *right now* is the IP address, then let's only work with that and make sure we're only passing the IP address into the closure, like so:
```

**Review by whatyouhide — APPROVED** [review]

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md (hunk 9 above)
Comment (Nezteb) [reply lib/elixir/pages/anti-patterns/process-anti-patterns.md:299]: ✍️Andrea✍️hates✍️Latin✍️ (I joke, I joke!)

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md (hunk 10 above)
Comment (PragTob) [reply lib/elixir/pages/anti-patterns/process-anti-patterns.md:316]: Is "as a closure" right here? I thought it was "within the closure of the spawned function" - as in, the function has a closure and storing the `conn` is part of it - the `conn` is not the closure itself :thinking: 

Adding now, we can rework later if needed - already provides more value, thank you!

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md (hunk 8 above)
Comment (PragTob) [reply lib/elixir/pages/anti-patterns/process-anti-patterns.md:298]: Thank you, long sentences are one of the banes/improvement points of my writing :'(

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md (hunk 10 above)
Comment (PragTob) [reply lib/elixir/pages/anti-patterns/process-anti-patterns.md]: (unresolved to clear the point about closures)

**Comment by PragTob** [comment]
And a big thank you to @whatyouhide !

IMG20230429163945

File: lib/elixir/pages/anti-patterns/process-anti-patterns.md (hunk 10 above)
Comment (whatyouhide) [reply lib/elixir/pages/anti-patterns/process-anti-patterns.md]: Your version is more technical but I wanted to try and keep it simple 😉

**Review by v0idpwn — COMMENTED** [review]
Looks great. I suggest also adding an example (either here or on GenServer docs) about doing this when sending messages. While it's way more obvious, it also happens :)

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by PragTob** [comment]
@josevalim thanks for the merge and as I don't think I'll open another issue/MR that soon: Happy holidays! :green_heart: 

IMG_20180603_163744
